tools/harness-simulation/posix/sniffer_sim/sniffer.py

```diff
@@ -28,104 +28,172 @@
+            self._mutex.release()
+
+            # Equivalent to RF enclosure
+            if nodeid in allowed_nodeids:
+                self._pcap.append(data)
 
         self.logger.debug('Sniffer stopped.')
 
-    def start(self):
+    def Start(self, request, context):
         """ Start sniffing. """
 
+        self.logger.debug('call Start')
+
+        # Validate and change the state
+        if self._state != SnifferServicer.State.STOPPED:
+            return sniffer_pb2.StatusReply(status=sniffer_pb2.OPERATION_ERROR)
+        self._state = SnifferServicer.State.RUNNING
+
+        self._pcap = pcap_codec.PcapCodec(request.channel)
+
+        # Sniffer all nodes in default, i.e. there is no RF enclosure
+        self._allowed_nodeids = set(range(1, self.MAX_NODES_NUM + 1))
+
+        # Create transport
+        transport_factory = sniffer_transport.SnifferTransportFactory()
+        self._transport = transport_factory.create_transport()
+
+        # Start the sniffer main loop thread
         self._thread = threading.Thread(target=self._sniffer_main_loop)
         self._thread.daemon = True
-
         self._transport.open()
-
         self._thread_alive.set()
         self._thread.start()
 
-    def stop(self):
-        """ Stop sniffing. """
+        return sniffer_pb2.StatusReply(status=sniffer_pb2.OK)
 
-        self._thread_alive.clear()
+    def FilterNodes(self, request_iterator, context):
+        """ Only sniffer the specified nodes. """
 
-        self._transport.close()
+        self.logger.debug('call FilterNodes')
+
+        # Validate the state
+        if self._state != SnifferServicer.State.RUNNING:
+            return sniffer_pb2.StatusReply(status=sniffer_pb2.OPERATION_ERROR)
+
+        allowed_nodeids = set([x.nodeid for x in request_iterator])
+        # Validate the node IDs
+        for nodeid in allowed_nodeids:
+            if not 1 <= nodeid <= self.MAX_NODES_NUM:
+                return sniffer_pb2.StatusReply(status=sniffer_pb2.VALUE_ERROR)
+
+        self._mutex.acquire()
+        self._allowed_nodeids = allowed_nodeids
+        self._mutex.release()
+
+        return sniffer_pb2.StatusReply(status=sniffer_pb2.OK)
 
+    def Stop(self, request, context):
+        """ Stop sniffing, and return the pcap bytes. """
+
+        self.logger.debug('call Stop')
+
+        # Validate and change the state
+        if self._state != SnifferServicer.State.RUNNING:
+            return sniffer_pb2.StopReply(status=sniffer_pb2.OPERATION_ERROR, pcap_content=b'')
+        self._state = SnifferServicer.State.STOPPED
+
+        self._thread_alive.clear()
         self._thread.join(timeout=1)
-        self._thread = None
+        self._transport.close()
+
+        pcap_content = self._pcap.pop_all()
+        self._clear()
+
+        return sniffer_pb2.StopReply(status=sniffer_pb2.OK, pcap_content=pcap_content)
+
+
+def serve(address_port):
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=8))
```

Comment:
```suggestion
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=1))
```
I guess 1 worker is enough.

Reply:
Accepted.